### PR TITLE
Implement it in another way

### PR DESCRIPTION
**Pull request by jvican**
Tips:

* Scope as much as possible. Control the scopes and check that every
  task you use inside your own tasks is correctly scoped by checking with
  `inspect` where it's scoped or if it's scoped at all.
* It's not necessary to manually detect which is the caller of our task
  (differentiate between which either compile or test:compile is running).
* Only use `ThisBuild` for settings that are true for every project and
  can be invoked from there.
* Separate implementation of *your* task and the definition of it in the
  projects. It's better to define the Scala implementation either in
  Scala files or in defs, and then invoke those with the correct inputs
  from the project you're scoping in.

I should put these tips somewhere. But I do think that sbt could do
better to detect keys badly scoped.

This could surely be improved. Please, have a look at the implementation and
tell me the things that are not obvious, and I'll see way to improve it.

### CODE REVIEW

**Comment by rbellamy** [comment]
I'm looking at it right now - for some reason I didn't see it until just now.

**Comment by rbellamy** [comment]
Huh... bad rebase. Sorry about that... testing it now.

**Comment by rbellamy** [comment]
It's almost there.
```
> show compile:dbConf
[info] DbConf(slick.jdbc.PostgresProfile$,org.postgresql.Driver,jdbc:postgresql://localhost:5432/test,postgres,admin123)
> show test:dbConf
[info] DbConf(slick.jdbc.PostgresProfile$,org.postgresql.Driver,jdbc:postgresql://localhost:5434/test,postgres,admin123)
> show compile:managedSources
[info] flyway/compile:managedSources
[info] 	List()
[info] flyway-slick-codegen/compile:managedSources
[info] 	List(/home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/src_managed/main/com/terradatum/dao/Tables.scala)
[success] Total time: 1 s, completed May 7, 2017 2:22:06 PM
> show test:managedSources
[info] flyway/test:managedSources
[info] 	List()
[info] flyway-slick-codegen/test:managedSources
[info] 	List(/home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/src_managed/test/com/terradatum/dao/Tables.scala)
[success] Total time: 0 s, completed May 7, 2017 2:22:09 PM
```

So the scoped configuration is working for the `dbConf` `settingKey` and for the scoped `sourceGenerators`.

However, note that the `test:compile` runs the code generation twice:
```
> test:compile
[info] Updating {file:/home/rbellamy/Development/flyway-slick-codegen/}flyway...
[info] Resolving jline#jline;2.12.1 ...
[info] Done updating.
[info] Updating {file:/home/rbellamy/Development/flyway-slick-codegen/}flyway-slick-codegen...
[info] Resolving jline#jline;2.12.1 ...
[info] Done updating.
[info] Running slick.codegen.SourceCodeGenerator slick.jdbc.PostgresProfile org.postgresql.Driver jdbc:postgresql://localhost:5432/test /home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/src_managed/main com.terradatum.dao postgres admin123
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[info] Compiling 3 Scala sources to /home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/classes...
[warn] there were three unchecked warnings; re-run with -unchecked for details
[warn] there was one feature warning; re-run with -feature for details
[warn] two warnings found
[info] Running slick.codegen.SourceCodeGenerator slick.jdbc.PostgresProfile org.postgresql.Driver jdbc:postgresql://localhost:5434/test /home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/src_managed/test com.terradatum.dao postgres admin123
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[info] Compiling 1 Scala source to /home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/test-classes...
```

This is because the `test:sourceGenerators` is targeting the `Test` `sourceManaged` directory.

**Comment by rbellamy** [comment]
@jvican thank you very much for looking at this! I see where you are going with things, I think...

For instance:
```
def generatorsSetting: Seq[Setting[_]] =
  sourceGenerators += Def.taskDyn(generateTables((dbConf in flyway).value, dependencyClasspath.value)).taskValue

def allSourceGenerators: Seq[Setting[_]] =
  inConfig(Compile)(generatorsSetting) ++ inConfig(Test)(generatorsSetting)
```
Something about the above just isn't gelling for me.

Specifically, if I understand what's happening there, when SBT is in the `Test` configuration scope, the `generatedTables` task will be passed in the equivalent of `dbConf in flyway in Test`, correct?

**Comment by dwijnand** [comment]
Yes, `inConfig` (and friends) will map the keys on the left-hand side (`sourceGenerators` in this case) as well as those referenced in right-hand side (`dbConf in flyway` and `dependencyClasspath` in this case) to be scoped within the configuration specified.

**Comment by rbellamy** [comment]
Where things are still not falling into place is that the code generated is being placed in the `sourceGenerators in flyway in Test`.

Regardless of which configuration is used (`Compile` or `Test`), the code should be generated into `scala/src_managed/main`.

```
> show managedSources
[info] Running slick.codegen.SourceCodeGenerator slick.jdbc.PostgresProfile org.postgresql.Driver jdbc:postgresql://localhost:5432/test /home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/src_managed/main com.terradatum.dao postgres admin123
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[info] flyway/compile:managedSources
[info] 	List()
[info] flyway-slick-codegen/compile:managedSources
[info] 	List(/home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/src_managed/main/com/terradatum/dao/Tables.scala)
[success] Total time: 3 s, completed May 8, 2017 6:53:11 AM
> show test:managedSources
[info] Compiling 3 Scala sources to /home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/classes...
[warn] there were three unchecked warnings; re-run with -unchecked for details
[warn] there was one feature warning; re-run with -feature for details
[warn] two warnings found
[info] Running slick.codegen.SourceCodeGenerator slick.jdbc.PostgresProfile org.postgresql.Driver jdbc:postgresql://localhost:5434/test /home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/src_managed/test com.terradatum.dao postgres admin123
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[info] flyway/test:managedSources
[info] 	List()
[info] flyway-slick-codegen/test:managedSources
[info] 	List(/home/rbellamy/Development/flyway-slick-codegen/target/scala-2.11/src_managed/test/com/terradatum/dao/Tables.scala)
[success] Total time: 11 s, completed May 8, 2017 6:53:27 AM
```

**Comment by dwijnand** [comment]
@rbellamy I pushed a change that should have fixed that.

Unfortunately I can't test it locally because I'm getting the following (and I don't have time to investigate the cause of it):

```
$ ./docker-create.sh
NOTICE:  database "test" does not exist, skipping
psql:drop-role.sql:1: ERROR:  role "test" does not exist
CREATE ROLE
createdb: database creation failed: FATAL:  terminating connection due to administrator command
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
psql: FATAL:  the database system is shutting down
```

File: build.sbt [hunk 1]
```diff
@@ -54,11 +54,14 @@ def generateTables(conf: DbConf, dependencyClasspath: Seq[Attributed[File]]) = D
   Seq(file(fname))
 }
 
-def generatorsSetting: Seq[Setting[_]] =
-  sourceGenerators += Def.taskDyn(generateTables((dbConf in flyway).value, dependencyClasspath.value)).taskValue
-
-def allSourceGenerators: Seq[Setting[_]] =
-  inConfig(Compile)(generatorsSetting) ++ inConfig(Test)(generatorsSetting)
+def allSourceGenerators: Seq[Setting[_]] = Seq(
```
Comment (jvican) [inline build.sbt]: :+1: Well done.

**Comment by dwijnand** [comment]
But ultimately what you're trying to do goes against a strong convention in sbt (derived from Maven, and probably others): that test depends on compile.

For example you want flyway and slick codegen to operate differently depending on if you are going to run the tests or not. But as a counter-example you're happy for `compile` to first compile your main sources and then use the result of that to compile your test sources.

The solution I recommend (echoing Paul's) is to use projects rather than configurations to solve this. So move any shared sources into a shared sources directory. Then create one project for when you want to target the runtime database, another project to target the test database, and make them share the sources. You'll compile your shared sources twice, but it would be much easier to setup your workflow.

**Comment by rbellamy** [comment]
@dwijnand I'm testing this right now. Thank you for your continued support and helpful hand.

I disagree that what I want is unconventional. I think SBT has forced my discussion through a particular set of cognitive filters and bias to the point where what I want only SEEMS unconventional.

What I want to do is use a test database for migration and code generation when running tests, and a runtime database otherwise. That is patently NOT unconventional. I've done it in Maven and Gradle often - otherwise, how, exactly, am I supposed to actually TEST my migrations and codegen during development?

**Comment by rbellamy** [comment]
This update runs migrations on `localhost:5434` and cogegen on `localhost:5432`

**Pull request merged**